### PR TITLE
fix broken REST API for the Rule / Automation API

### LINKED ISSUE #1801: Cannot create rules using Paper UI / configuration broken
My intention was to test #1683 but it seems it is not possible for me to use the trigger to create rules using the Paper UI.

I added two triggers (two item state change triggers on the first try, two item state update triggers on the second try) to the rule and tried to save the rule.
I see an error on the console log.

I have used the Firefox to create that rules.

```
git rev-parse HEAD
8aea602d7adee12e5008ef065d27adad8337d8eb
```

```text
22:01:32.790 INFO  smarthome.event.RuleAddedEvent[:43] - Rule 'rule_1' has been added.
22:01:32.790 DEBUG o.e.s.a.c.internal.RuleEngine[:219] - Added rule 'rule_1'
22:01:32.795 INFO  s.event.RuleStatusInfoEvent[:43] - rule_1 updated: NOT_INITIALIZED
22:01:32.795 DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory[:101] - create composite module:trigger_1, of rule: rule_1
22:01:32.798 WARN  o.e.s.a.c.i.ReferenceResolverUtil[:232] - Cannot find reference for ${ itemName } , it will remain the same.
22:01:32.806 DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory[:132] - Set handler to child module: org.eclipse.smarthome.automation.core.internal.RuntimeTrigger@2b769047  -> CompositeTriggerHandler.
22:01:32.808 DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory[:101] - create composite module:trigger_2, of rule: rule_1
22:01:32.809 WARN  o.e.s.a.c.i.ReferenceResolverUtil[:232] - Cannot find reference for ${ itemName } , it will remain the same.
22:01:32.810 DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory[:132] - Set handler to child module: org.eclipse.smarthome.automation.core.internal.RuntimeTrigger@4b13be54  -> CompositeTriggerHandler.
22:01:32.829 ERROR o.e.s.a.c.i.RuleRegistryImpl[:236] - Can't add rule: rule_1
java.lang.RuntimeException: Missing required parameter: itemName of type ItemStateChangeTrigger
	at org.eclipse.smarthome.automation.core.internal.RuleEngine.setDefaultConfigurationValues(RuleEngine.java:1290)
	at org.eclipse.smarthome.automation.core.internal.RuleEngine.setRule(RuleEngine.java:356)
	at org.eclipse.smarthome.automation.core.internal.RuleEngine.setRuleEnabled(RuleEngine.java:702)
	at org.eclipse.smarthome.automation.core.internal.RuleEngine.addRule0(RuleEngine.java:221)
	at org.eclipse.smarthome.automation.core.internal.RuleEngine.addRule(RuleEngine.java:201)
	at org.eclipse.smarthome.automation.core.internal.RuleRegistryImpl.onAddElement(RuleRegistryImpl.java:231)
	at org.eclipse.smarthome.automation.core.internal.RuleRegistryImpl.onAddElement(RuleRegistryImpl.java:1)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.added(AbstractRegistry.java:56)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:51)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:70)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListenersAboutAddedElement(AbstractProvider.java:74)
22:01:32.830 DEBUG o.e.s.c.c.r.AbstractManagedProvider[:62] - Added new element rule_1 to ManagedRuleProvider.
```

```text
22:03:41.515 DEBUG o.e.s.a.c.internal.RuleEngine[:310] - Updated rule 'rule_1'.
22:03:41.516 DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory[:101] - create composite module:trigger_4, of rule: rule_1
22:03:41.518 DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory[:132] - Set handler to child module: org.eclipse.smarthome.automation.core.internal.RuntimeTrigger@35269352  -> CompositeTriggerHandler.
22:03:41.519 DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory[:101] - create composite module:trigger_5, of rule: rule_1
22:03:41.520 DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory[:132] - Set handler to child module: org.eclipse.smarthome.automation.core.internal.RuntimeTrigger@211e61a4  -> CompositeTriggerHandler.
22:03:41.523 WARN  o.e.s.c.c.r.AbstractRegistry[:104] - Could not update element: Missing required parameter: itemName of type ItemStateUpdateTrigger
java.lang.RuntimeException: Missing required parameter: itemName of type ItemStateUpdateTrigger
	at org.eclipse.smarthome.automation.core.internal.RuleEngine.setDefaultConfigurationValues(RuleEngine.java:1290)
	at org.eclipse.smarthome.automation.core.internal.RuleEngine.setRule(RuleEngine.java:356)
	at org.eclipse.smarthome.automation.core.internal.RuleEngine.updateRule(RuleEngine.java:318)
	at org.eclipse.smarthome.automation.core.internal.RuleRegistryImpl.onUpdateElement(RuleRegistryImpl.java:291)
	at org.eclipse.smarthome.automation.core.internal.RuleRegistryImpl.onUpdateElement(RuleRegistryImpl.java:1)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.updated(AbstractRegistry.java:99)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:57)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListenersAboutUpdatedElement(AbstractProvider.java:82)
	at org.eclipse.smarthome.core.common.registry.AbstractManagedProvider.update(AbstractManagedProvider.java:134)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.update(AbstractRegistry.java:122)
	at org.eclipse.smarthome.automation.rest.internal.RuleResource.update(RuleResource.java:163)
22:03:41.524 DEBUG o.e.s.c.c.r.AbstractManagedProvider[:135] - Updated element rule_1 in ManagedRuleProvider.
```

### PR DESCRIPTION
**Pull request by maggu2810**
This should fix #1801

@danchom Can you please test this changes, too?

### CODE REVIEW

**Comment by maggu2810** [comment]
I think this fixes the Rule REST API.

There is also the module-type REST API.
* A ModuleType could be a TriggerType.
* A TriggerType could be a CompositeTriggerType
* A CompositeTriggerType has a List of children of type Trigger
* Tigger contains a configuration, so a TriggerDTO have to be used
* => a CompositeTriggerTypeDTO have to be used
* => a TriggerTypeDTO have to be used
* => a ModuleType trigger type have to be used

If we are not interested in the members of the sub-types of a TriggerType (e.g. the children member of the composite trigger type - same for other module-types), then it should be enough to create a
* ModuleTypeDTO
* ActionTypeDTO
* ConditionTypeDTO
* TriggerTypeDTO

WDYT @kaikreuzer 

So, I assume we need

**Comment by danchom** [comment]
I think we should access, through the REST API, the children modules of composite module type. If it is not needed I don't see a reason to define ModuleTypeDTO, ActionTypeDTO, ConditionTypeDTO, TriggerTypeDTO at all.

**Comment by maggu2810** [comment]
I think we should use the DTOs and not transport the internal representation of every module type (e.g. a specific trigger type) to the outside world.

If I create a trigger type that has a lot of private members, I do not want to give all that stuff to the user of the REST API.
The REST API should give us the available trigger / action / condition type and we should handle a trigger type by the information a trigger type provides.
If a special trigger type is implemented to use children or whatever should be out of scope.
IMHO that is internal representation of the type implementation.

We want to provide the trigger type and that interface should be stable, the implementation should be kept internal.

Why is a DTO not needed? How do you want to ensure that no private information is made public?

**Comment by danchom** [comment]
"Why is a DTO not needed? How do you want to ensure that no private information is made public?" 
There is no private information in returned public module type object. The ModuleType registry always returns a copy of the object which is used by it. And creation of DTO is one more object copy. Also I think the Java API must be close to the REST API, when we have different object presentation (ModuleType and ModuleTypeDTO) this relation is not so obvious. Of course when this is needed (EnrichedRuleDTO, ModuleDTO)  it has to be done. But if you think this is clear architecture I'm ok with that.
Composite module types are provided and handled by the automation.core and for that reason I think the ModuleTypeDTO must have children property containing child modules. Otherwise the rest api  will not support composite module types. 
For example the ItemStateUpdateTrigger is a composite module.
Besides this the PR LGTM and it fixes the configuration problem in REST API

**Comment by maggu2810** [comment]
> There is no private information in returned public module type object.

Without the DTOs the REST API returns the whole members of the respective module type.

Using a CompositeTriggerType the REST API offers all the children triggers with all their configuration.
At least here also the configuration is returned with its member "properties" and all the stuff that is contained in the configuration properties.
This and the other PR should hide that the Configuration contains a member "properties" that is a map and the Configuration should be a key value pair collection at all.

Please have a look at the result of the REST API call, at least on my machines I see much more information then is defined in the TriggerType class.
I see all members of the specific implementation of that class.

> Composite module types are provided and handled by the automation.core and for that reason I think the ModuleTypeDTO must have children property containing child modules. Otherwise the rest api will not support composite module types. 

The REST API should (but I could be wrong) give a list of all available module-types (condition, action, trigger).
It should not mirror the implementation. If a condition type is internally implemented using composite or not composite should the user not care about.
I implemented the Item State Update Trigger using no composite type and I have the one using the composite type.

Why should one stop working, if the one that asks for trigger type is not informed how it is implemented?

A trigger type is defined by the trigger type class and the DTO ensures that only that information are made public through the REST API. No more information should be necessary.



But I am not as deep as you in the internals of all that automation stuff.

**Comment by danchom** [comment]
"The REST API should (but I could be wrong) give a list of all available module-types (condition, action, trigger).
It should not mirror the implementation. If a condition type is internally implemented using composite or not composite should the user not care about."
OK I agree

**Comment by kaikreuzer** [comment]
@maggu2810: Designing the rule engine, we had tried very hard to avoid having separate DTOs - the goal was to have the Rule and Action/Trigger/Condition classes be clean Pojos that could be used "as-is" in the REST API and anywhere else.

I agree that we seem to have missed that goal, which becomes obvious with the serialization of the Configuration and with the exposed internals of the composite implementations. So although it is a pity, I think we probably won't come around DTOs anymore.

What I would suggest though, is to follow the same structure as we have for other DTOs:

- They are NOT part of the REST bundle, but available in a subpackage of the API itself (see e.g. org.eclipse.smarthome.core.items.dto. We put them there (after having them originally within the rest bundle), because they are also used for serialization in events (ItemAddedEvent) and possibly in other situations.
- Only the EnrichedDTOs are specifically designed for the REST API and should stay in that bundle.
- The DTOs are Pojos without any additional methods/functionality (like your toDtoList/fromDtoList/createXXX). Mapping functionality is instead provided in a separate "Mapper" class (e.g. ItemDTOMapper). I don't say that this is better than your approach, I just think we should follow the same pattern in all places.

**Comment by maggu2810** [comment]
> I don't say that this is better than your approach, I just think we should follow the same pattern in all places.

Okay for me.

I will implement the changes the next days.

**Comment by maggu2810** [comment]
Generics are not available at Runtime, so using equal method names for e.g. `List<ActionDTO> map(List<Action> list)` and `List<Action> map(List<ActionDTO> list)` does not work.
That is the reason, why I named some methods mapDto instead of map.

Using the separate mapper classes keeps the POJOs as simple as them could be but adds more lines of code and the DTO class and the one that fills it are not in the same file.
So if you touch the POJO, you must also _know_ the file that fills it.
As stated above, okay for consistence but that is the reason the PR is a little big for the stuff it handles... 

I hope there is nothing missing.

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/dto/ActionDTO.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/**
+ * Copyright (c) 2016 Markus Rathgeb
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/dto/ActionDTO.java]: could you please add "and others"?

**Comment by kaikreuzer** [comment]
Thanks @maggu2810, looks good to me. @danchom, any comments from your side?

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/dto/ActionDTO.java (hunk 1 above)
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/dto/ActionDTO.java]: Sure, will do that for all copyright headers of new files of this PR.

FYI (why it has been done without others) : As I don't know how to do it correctly I have read 
https://eclipse.org/legal/copyrightandlicensenotice.php

<pre>If the content is subsequently modified and appended to by other copyright owners, the words "and others" are typically appended.</pre>

I assumed this is done on the first modification.
Will do it on creation from now and change this PR.

Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/dto/ActionDTO.java]: As discussed in the past, the agreed procedure in this project is [A.2 from the guidelines](https://www.eclipse.org/smarthome/documentation/development/guidelines.html) and keeping track of the copyrights through A.4 (author tags). I would actually like to change the automatically generated headers to be [similar to this](https://github.com/openhab/openhab2-addons/blob/master/src/etc/header.txt#L1). Wdyt?

**Comment by danchom** [comment]
LGTM

**Pull request merged**